NeoML/test/src/ReferenceDnnTest.cpp

```diff
@@ -0,0 +1,279 @@
+	fc1->SetNumberOfElements(50);
+	fc1->SetName("fc1");
+	fc1->Connect(0, *dataLayer);
+	net->AddLayer(*fc1);
+
+	CPtr<CDropoutLayer> dp1 = new CDropoutLayer(MathEngine());
+	dp1->SetName("dp1");
+	dp1->SetDropoutRate(0.1f);
+	dp1->Connect(0, *fc1);
+	net->AddLayer(*dp1);
+
+	CPtr<CFullyConnectedLayer> fc2 = new CFullyConnectedLayer(MathEngine());
+	fc2->SetName("fc2");
+	fc2->SetNumberOfElements(20);
+	fc2->Connect(*dp1);
+	net->AddLayer(*fc2);
+
+	CPtr<CDropoutLayer> dp2 = new CDropoutLayer(MathEngine());
+	dp2->SetName("dp2");
+	dp2->SetDropoutRate(0.1f);
+	dp2->Connect(0, *fc2);
+	net->AddLayer(*dp2);
+
+	CPtr<CFullyConnectedLayer> fc3 = new CFullyConnectedLayer(MathEngine());
+	fc3->SetName("fc3");
+	fc3->SetNumberOfElements(10);
+	fc3->Connect(0, *dp2);
+	net->AddLayer(*fc3);
+
+	CPtr<CSinkLayer> sink = new CSinkLayer(MathEngine());
+	sink->SetName("sink");
+	sink->Connect(0, *fc3);
+	net->AddLayer(*sink);
+
+	return net;
+}
+
+} // namespace NeoMLTest
+
+//------------------------------------------------------------------------------------------------
+
+TEST(ReferenceDnnTest, ReferenceDnnsThreads)
+{
+	const int numOfThreads = 4;
+
+	CObjectArray<CDnnBlob> blobs;
+	CArray<CDnnTestParam> taskParams;
+	CArray<CRandom> randoms;
+	CArray<CDnn*> dnns;
+
+	IPerformanceCounters* counters(MathEngine().CreatePerformanceCounters());
+	counters->Synchronise();
+
+	for(int i = 0; i < numOfThreads; ++i) {
+		randoms.Add(CRandom(i));
+		if(i == 0) {
+			dnns.Add(createDnn(randoms[i]));
+		} else {
+			dnns.Add(dnns[0]->CreateReferenceDnn(randoms[i]));
+		}
+		blobs.Add(CDnnBlob::CreateTensor(MathEngine(), CT_Float, { 1, 1, 1, 8, 20, 30, 10 }));
+		taskParams.Add({ dnns[i] });
+		
+		CRandom randomInit(0);
+		for(int j = 0; j < blobs[i]->GetDataSize(); ++j) {
+			blobs[i]->GetData().SetValueAt(j, static_cast<float>(randomInit.Uniform(-1, 1)));
+		}
+		static_cast<CSourceLayer*>(dnns[i]->GetLayer("in").Ptr())->SetBlob(blobs[i]);
+	}
+
+	IThreadPool* pool = CreateThreadPool(numOfThreads);
+	for(int i = 0; i < numOfThreads; ++i) {
+		pool->AddTask(i, runDnn, &(taskParams[i]));
+	}
+
+	pool->WaitAllTask();
+	counters->Synchronise();
+	std::cerr
+		<< '\n' << "Time: " << (double((*counters)[0].Value) / 1000000) << " ms."
+		<< '\t' << "Peak.Mem: " << (double(MathEngine().GetPeakMemoryUsage()) / 1024 / 1024) << " MB \n";
+
+	for(int i = 0; i < numOfThreads - 1; ++i) {
+		EXPECT_TRUE(CompareBlobs(
+			*(static_cast<CSourceLayer*>(dnns[i]->GetLayer("in").Ptr())->GetBlob()),
+			*(static_cast<CSourceLayer*>(dnns[i + 1]->GetLayer("in").Ptr())->GetBlob()))
+		);
+
+		EXPECT_TRUE(CompareBlobs(
+			*(static_cast<CSinkLayer*>(dnns[i]->GetLayer("sink").Ptr())->GetBlob()),
+			*(static_cast<CSinkLayer*>(dnns[i+1]->GetLayer("sink").Ptr())->GetBlob()))
+		);
+	}
+
+	delete counters;
+	delete pool;
+	for(int i = 0; i < numOfThreads; ++i) {
+		delete dnns[i];
+	}
+}
+
+TEST(ReferenceDnnTest, FullCopyDnnsThreads)
```

Comment:
Can you unite the common part of these 3 tests?
It better to see for the performance-tests that code is the same except a small part, but the results are different.

Please, add 
* 2 check-consistance tests of CReferenceDnn 
* 2 performance tests CReferenceDnn/CDnn

performance tests should be disabled by default, like in loraTest
DISABLED_<TestName>